src/asymmetric_crypto/rsa/private_key.rs

```diff
@@ -0,0 +1,151 @@
+use digest::{Digest, DynDigest};
+use pkcs8::SecretDocument;
+use rand_chacha::rand_core::CryptoRngCore;
+use rand_core::RngCore;
+use rsa::traits::PublicKeyParts;
+use zeroize::{ZeroizeOnDrop, Zeroizing};
+
+use crate::{
+    key_unwrap, pkcs8_fix, CryptoCoreError, PrivateKey, RsaKeyLength, RsaKeyWrappingAlgorithm,
+    RsaPublicKey,
+};
+
+#[derive(Hash, Clone, Debug, PartialEq, Eq)]
+pub struct RsaPrivateKey(rsa::RsaPrivateKey);
+
+impl RsaPrivateKey {
+    /// Generate a random private key
+    pub fn new<R: CryptoRngCore>(
+        rng: &mut R,
+        key_length: RsaKeyLength,
+    ) -> Result<Self, CryptoCoreError> {
+        Ok(Self(rsa::RsaPrivateKey::new(
+            rng,
+            (key_length as i32) as usize,
+        )?))
+    }
+
+    /// Get the key length which is the modulus size in bits
+    pub fn key_length(&self) -> RsaKeyLength {
+        match self.0.n().bits() {
+            2048 => RsaKeyLength::Modulus2048,
+            3072 => RsaKeyLength::Modulus3072,
+            4096 => RsaKeyLength::Modulus4096,
+            _ => panic!("Invalid RSA key length; this should never happen"),
+        }
+    }
```

Comment:
You could also have stored the key inside an `enum`:
```rust
pub enum RsaPrivateKey {
    Modulus2048(rsa::RsaPrivateKey),
    ...
}
```
Is there an issue with this?

Reply:
This makes the high-level RsaPrivateKey object an enum, rather than a struct, which is a completely different approach. I guess this would work too.  The current implementation mirrors the implementation of the rsa crate and the ones we have for symmetric keys where then length is a property